File: DataEngineering/06_Data_Orchestration/projects/golang/04_data_pipeline_monitor/main.go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var (
	pipelineDuration = prometheus.NewSummaryVec(
		prometheus.SummaryOpts{
			Name: "pipeline_processing_seconds",
			Help: "Time spent processing pipeline tasks",
		},
		[]string{"task_name"},
	)
	pipelineStatus = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "pipeline_task_status_total",
			Help: "Count of pipeline task statuses",
		},
		[]string{"task_name", "status"},
	)
)

func init() {
	// Register metrics with Prometheus's default registry.
	prometheus.MustRegister(pipelineDuration)
	prometheus.MustRegister(pipelineStatus)
}

func executeTask(taskName string) error {
	log.Printf("Executing task: %s\n", taskName)

	start := time.Now()
	defer func() {
		pipelineDuration.WithLabelValues(taskName).Observe(time.Since(start).Seconds())
	}()

	// Simulate processing time
	time.Sleep(time.Duration(500+rand.Intn(1500)) * time.Millisecond)

	// Simulate occasional failures
	if rand.Float32() < 0.2 {
		pipelineStatus.WithLabelValues(taskName, "failed").Inc()
		log.Printf("Task failed: %s\n", taskName)
		return fmt.Errorf("task %s encountered an error", taskName)
	}

	pipelineStatus.WithLabelValues(taskName, "success").Inc()
	log.Printf("Task completed: %s\n", taskName)
	return nil
}

func runPipeline() {
	tasks := []string{"Extract", "Transform", "Load"}
	for _, task := range tasks {
		err := executeTask(task)
		if err != nil {
			log.Printf("Pipeline halted due to error: %v\n", err)
			break
		}
	}
	log.Println("Pipeline run finished.")
}

func main() {
	rand.Seed(time.Now().UnixNano())

	// Expose the registered metrics via HTTP.
	// Data Engineers use Prometheus + Grafana to scrape these metrics
	http.Handle("/metrics", promhttp.Handler())
	
	// Start metrics server in a goroutine
	go func() {
		log.Println("Starting Prometheus metrics server on :8000")
		if err := http.ListenAndServe(":8000", nil); err != nil {
			log.Fatalf("Error starting HTTP server: %v", err)
		}
	}()

	// Continuously run pipeline to generate metrics
	for {
		runPipeline()
		log.Println("Sleeping before next pipeline run...")
		time.Sleep(5 * time.Second)
	}
}

File: DataEngineering/06_Data_Orchestration/projects/python/04_data_pipeline_monitor/main.py
import time
import random
import logging
from prometheus_client import start_http_server, Summary, Counter

# Configure logging
logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
logger = logging.getLogger(__name__)

# Define Prometheus metrics
PIPELINE_DURATION = Summary('pipeline_processing_seconds', 'Time spent processing pipeline tasks', ['task_name'])
PIPELINE_STATUS = Counter('pipeline_task_status_total', 'Count of pipeline task statuses', ['task_name', 'status'])

class DataPipelineException(Exception):
    pass

def execute_task(task_name: str):
    logger.info(f"Executing task: {task_name}")
    
    # Track the duration of the task
    with PIPELINE_DURATION.labels(task_name=task_name).time():
        # Simulate processing time
        time.sleep(random.uniform(0.5, 2.0))
        
        # Simulate occasional failures for monitoring demonstration
        if random.random() < 0.2:
            PIPELINE_STATUS.labels(task_name=task_name, status="failed").inc()
            logger.error(f"Task failed: {task_name}")
            raise DataPipelineException(f"Task {task_name} encountered an error.")
            
        PIPELINE_STATUS.labels(task_name=task_name, status="success").inc()
        logger.info(f"Task completed: {task_name}")

def run_pipeline():
    tasks = ["Extract", "Transform", "Load"]
    for task in tasks:
        try:
            execute_task(task)
        except DataPipelineException as e:
            logger.critical(f"Pipeline halted due to error: {e}")
            break
    logger.info("Pipeline run finished.")

if __name__ == "__main__":
    # Start the Prometheus metrics HTTP server on port 8000
    # Data Engineers use Prometheus + Grafana to scrape these metrics and build dashboards
    logger.info("Starting Prometheus metrics server on port 8000")
    start_http_server(8000)
    
    # Run the pipeline periodically to generate metrics continuously
    try:
        while True:
            run_pipeline()
            logger.info("Sleeping before next pipeline run...")
            time.sleep(5)
    except KeyboardInterrupt:
        logger.info("Shutting down data pipeline monitor.")
